server/src/main/java/org/elasticsearch/index/mapper/DataStreamTimestampFieldMapper.java

```diff
@@ -0,0 +1,189 @@
+        }
+
+        @Override
+        public Query termQuery(Object value, QueryShardContext context) {
+            throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] does not support term queries");
+        }
+
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] does not support exists queries");
+        }
+
+    }
+
+    public static class Builder extends MetadataFieldMapper.Builder<Builder> {
+
+        private String fieldName;
+
+        public Builder() {
+            super(NAME, Defaults.DATA_STREAM_TIMESTAMP_FIELD_TYPE);
+        }
+
+        public void setFieldName(String fieldName) {
+            this.fieldName = fieldName;
+        }
+
+        @Override
+        public MetadataFieldMapper build(BuilderContext context) {
+            return new DataStreamTimestampFieldMapper(
+                fieldType,
+                new DataStreamTimestampFieldFieldType(),
+                fieldName
+            );
+        }
+    }
+
+    public static class TypeParser implements MetadataFieldMapper.TypeParser {
+
+        @Override
+        public MetadataFieldMapper.Builder<?> parse(String name,
+                                                    Map<String, Object> node,
+                                                    ParserContext parserContext) throws MapperParsingException {
+            Builder builder = new Builder();
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
+                Map.Entry<String, Object> entry = iterator.next();
+                String fieldName = entry.getKey();
+                Object fieldNode = entry.getValue();
+                if (fieldName.equals("field_name")) {
+                    builder.setFieldName((String) fieldNode);
+                    iterator.remove();
+                }
+            }
+            return builder;
+        }
+
+        @Override
+        public MetadataFieldMapper getDefault(ParserContext parserContext) {
+            return new DataStreamTimestampFieldMapper(Defaults.DATA_STREAM_TIMESTAMP_FIELD_TYPE,
+                new DataStreamTimestampFieldFieldType(), null);
+        }
+    }
+
+    private final String fieldName;
+
+    private DataStreamTimestampFieldMapper(FieldType fieldType, MappedFieldType mappedFieldType, String fieldName) {
+        super(fieldType, mappedFieldType);
+        this.fieldName = fieldName;
+    }
+
+    @Override
+    public void preParse(ParseContext context) throws IOException {
+
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+
+    }
+
+    @Override
+    public void postParse(ParseContext context) throws IOException {
+        if (fieldName == null) {
+            // not configured, so skip the validation
+            return;
+        }
+
```

Comment:
It would be nice to add validation in the `MapperService` that the target field is of  type `date` and that it is indexed and has doc values. I  know that you validate this in the data stream code but it would be cleaner/safer to do it directly in the MapperService since the metadata field could be used outside of data streams ?

Reply:
> do it directly in the MapperService since the metadata field could be used outside of data streams ?

If this metadata field mapper has value outside data streams then yes it make sense to add the validation in `MapperService`.

Reply:
I like the idea of moving the validation down into the mapping code even if the field mapper isn't used outside of data streams. It's really nice to have all mapping validation embedded in the mappers/ mapping logic instead of some living outside.